File: ios/RunFormCoachAI/RunSessionManager.swift
import Foundation
import os.log

// MARK: - RunSessionManager

/// Orchestrates the real-time running coach pipeline:
/// `CoreMotionManager → CadenceDetector → GaitAnalyzer → AudioCoachEngine`
///
/// Manages session lifecycle with a state machine:
/// `idle → ready → running → paused → stopped`
///
/// Provides aggregated real-time data callbacks for UI updates.
///
/// Usage:
/// ```swift
/// let session = RunSessionManager(config: RunSessionConfig())
/// session.onMetricsUpdate = { cadence, gait in
///     // update UI
/// }
/// session.start()
/// // ... run ...
/// session.pause()
/// session.resume()
/// session.stop()
/// ```
public final class RunSessionManager: @unchecked Sendable {

    // MARK: - Public types

    /// Aggregated session metrics delivered to UI callbacks.
    public struct SessionMetrics: Sendable {
        public let cadence: CadenceSample?
        public let gait: GaitSnapshot?
        public let elapsedSeconds: TimeInterval
        public let state: RunSessionState
        public let promptHistory: [CoachPrompt]

        public init(
            cadence: CadenceSample? = nil,
            gait: GaitSnapshot? = nil,
            elapsedSeconds: TimeInterval = 0,
            state: RunSessionState = .idle,
            promptHistory: [CoachPrompt] = []
        ) {
            self.cadence = cadence
            self.gait = gait
            self.elapsedSeconds = elapsedSeconds
            self.state = state
            self.promptHistory = promptHistory
        }
    }

    // MARK: - Public properties

    /// Current session state.
    public private(set) var state: RunSessionState = .idle

    /// Session configuration.
    public let config: RunSessionConfig

    /// Elapsed session time in seconds.
    public private(set) var elapsedSeconds: TimeInterval = 0

    /// History of coaching prompts delivered this session.
    public private(set) var promptHistory: [CoachPrompt] = []

    /// Callback invoked on each metrics update (~1 Hz).
    public var onMetricsUpdate: (@Sendable (SessionMetrics) -> Void)?

    /// Callback invoked on state transitions.
    public var onStateChange: (@Sendable (RunSessionState, RunSessionState) -> Void)?

    /// Callback invoked when a coach prompt is spoken (for UI overlay).
    public var onCoachPrompt: (@Sendable (CoachPrompt) -> Void)?

    // MARK: - Pipeline components

    public let motionManager: CoreMotionManager
    public let cadenceDetector: CadenceDetector
    public let gaitAnalyzer: GaitAnalyzer
    public let audioCoach: AudioCoachEngine

    // MARK: - Private

    private let queue = DispatchQueue(label: "com.runformcoachai.session")
    private var sessionStartDate: Date?
    private var pauseStartDate: Date?
    private var totalPausedSeconds: TimeInterval = 0
    private var metricsTimer: DispatchSourceTimer?

    // MARK: - Init

    /// - Parameters:
    ///   - config: Session configuration (cadence target, language, intervals, etc.).
    ///   - motionManager: Optional pre-configured CoreMotionManager. Created if nil.
    ///   - cadenceDetector: Optional pre-configured CadenceDetector. Created if nil.
    ///   - gaitAnalyzer: Optional pre-configured GaitAnalyzer. Created if nil.
    ///   - audioCoach: Optional pre-configured AudioCoachEngine. Created if nil.
    public init(
        config: RunSessionConfig = RunSessionConfig(),
        motionManager: CoreMotionManager? = nil,
        cadenceDetector: CadenceDetector? = nil,
        gaitAnalyzer: GaitAnalyzer? = nil,
        audioCoach: AudioCoachEngine? = nil
    ) {
        self.config = config

        self.motionManager = motionManager ?? CoreMotionManager(
            samplingRate: config.samplingRate,
            windowSeconds: config.sensorWindowSeconds
        )

        self.cadenceDetector = cadenceDetector ?? CadenceDetector(
            alpha: 0.8,
            windowSeconds: config.sensorWindowSeconds
        )
        // Sync samplingRate for accurate timestamp estimation
        self.cadenceDetector.samplingRate = self.motionManager.samplingRate

        self.gaitAnalyzer = gaitAnalyzer ?? GaitAnalyzer(
            windowSeconds: config.sensorWindowSeconds,
            samplingRate: config.samplingRate
        )

        self.audioCoach = audioCoach ?? AudioCoachEngine(
            language: config.coachingLanguage,
            minInterval: config.minPromptInterval
        )

        setupPipeline()
    }

    // MARK: - Public API: State machine

    /// Transition from idle → ready → running.
    public func start() {
        queue.async { [weak self] in
            guard let self else { return }
            guard self.state == .idle || self.state == .ready else {
                os_log(.info, "RunSessionManager: cannot start from state %{public}@",
                       self.state.rawValue)
                return
            }
            self.transition(to: .ready)
            self.transition(to: .running)
            self.sessionStartDate = Date()
            self.totalPausedSeconds = 0
            self.elapsedSeconds = 0
            self.promptHistory.removeAll()
            self.motionManager.startUpdates()
            self.startMetricsTimer()
        }
    }

    /// Pause the session (sensors keep running but metrics and coaching pause).
    public func pause() {
        queue.async { [weak self] in
            guard let self, self.state == .running else { return }
            self.pauseStartDate = Date()
            self.transition(to: .paused)
        }
    }

    /// Resume from paused state.
    public func resume() {
        queue.async { [weak self] in
            guard let self, self.state == .paused else { return }
            if let pauseStart = self.pauseStartDate {
                self.totalPausedSeconds += Date().timeIntervalSince(pauseStart)
                self.pauseStartDate = nil
            }
            self.transition(to: .running)
        }
    }

    /// Stop the session completely.
    public func stop() {
        queue.async { [weak self] in
            guard let self, self.state != .idle, self.state != .stopped else { return }
            // Accumulate any active pause before stopping
            if let pauseStart = self.pauseStartDate {
                self.totalPausedSeconds += Date().timeIntervalSince(pauseStart)
                self.pauseStartDate = nil
            }
            self.motionManager.stopUpdates()
            self.stopMetricsTimer()
            self.cadenceDetector.reset()
            self.gaitAnalyzer.reset()
            self.audioCoach.stopSpeaking()
            self.transition(to: .stopped)
            self.transition(to: .idle)
        }
    }

    // MARK: - Pipeline setup

    private func setupPipeline() {
        // Wire: CoreMotion → CadenceDetector + GaitAnalyzer

        motionManager.onFrame = { [weak self] frame in
            guard let self else { return }
            // Feed cadence detector: use vertical acceleration magnitude
            let accelMag = sqrt(
                frame.accelerationX * frame.accelerationX +
                frame.accelerationY * frame.accelerationY +
                frame.accelerationZ * frame.accelerationZ
            )
            self.cadenceDetector.process(value: accelMag)

            // Feed gait analyzer with full frame
            self.gaitAnalyzer.process(frame: frame)
        }

        // Wire: CadenceDetector → AudioCoach evaluation
        cadenceDetector.onCadenceUpdate = { [weak self] cadence in
            guard let self else { return }
            // Inject latest cadence into gait analyzer for snapshot enrichment
            self.gaitAnalyzer.latestCadenceSPM = cadence.currentSPM
            let gait = self.gaitAnalyzer.currentSnapshot
            let elapsed = self.elapsedSeconds
            self.audioCoach.evaluate(
                cadence: cadence,
                gait: gait,
                targetCadence: self.config.targetCadenceSPM,
                elapsedSeconds: elapsed
            )
        }

        // Wire: AudioCoach prompts → UI callback
        audioCoach.onPromptQueued = { [weak self] prompt in
            guard let self else { return }
            self.promptHistory.append(prompt)
            // Keep last 50 prompts max
            if self.promptHistory.count > 50 {
                self.promptHistory.removeFirst(self.promptHistory.count - 50)
            }
            self.onCoachPrompt?(prompt)
        }
    }

    // MARK: - Metrics timer

    /// Fires at ~1 Hz to deliver aggregated session metrics to the UI.
    private func startMetricsTimer() {
        let timer = DispatchSource.makeTimerSource(queue: queue)
        timer.schedule(deadline: .now() + 1.0, repeating: 1.0)
        timer.setEventHandler { [weak self] in
            guard let self, self.state == .running else { return }
            if let start = self.sessionStartDate {
                self.elapsedSeconds = Date().timeIntervalSince(start) - self.totalPausedSeconds
            }
            let metrics = SessionMetrics(
                cadence: self.cadenceDetector.currentCadence,
                gait: self.gaitAnalyzer.currentSnapshot,
                elapsedSeconds: self.elapsedSeconds,
                state: self.state,
                promptHistory: self.promptHistory
            )
            self.onMetricsUpdate?(metrics)
        }
        timer.resume()
        metricsTimer = timer
    }

    private func stopMetricsTimer() {
        metricsTimer?.cancel()
        metricsTimer = nil
    }

    // MARK: - State transitions

    private func transition(to newState: RunSessionState) {
        let oldState = state
        guard oldState != newState else { return }
        state = newState
        os_log(.info, "RunSessionManager: %{public}@ → %{public}@",
               oldState.rawValue, newState.rawValue)
        onStateChange?(oldState, newState)
    }
}

File: android/app/src/main/java/com/runformcoach/runformcoachai/sensor/RunSessionManager.kt
package com.runformcoach.runformcoachai.sensor

import android.content.Context
import android.util.Log
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import javax.inject.Inject
import javax.inject.Singleton

// ── 会话状态枚举 ─────────────────────────────────────────────────────────────

/**
 * [RunSessionManager] 会话状态机状态。
 */
enum class RunSessionState {
    idle,
    ready,
    running,
    paused,
    stopped
}

// ── 会话配置 ──────────────────────────────────────────────────────────────────

/**
 * 实时跑步会话配置。
 *
 * @property targetCadenceSPM      目标步频 (步/分)
 * @property coachingLanguage      教练语言代码 ("en", "zh", "nl")
 * @property minPromptIntervalSec  语音提示最小间隔 (秒)
 * @property sensorWindowSeconds   传感器缓冲窗口 (秒), 限制 3–10
 * @property samplingRate          传感器采样率 (Hz)
 */
data class RunSessionConfig(
    val targetCadenceSPM: Double = 170.0,
    val coachingLanguage: String = "en",
    val minPromptIntervalSec: Int = 15,
    val sensorWindowSeconds: Double = 6.0,
    val samplingRate: Double = 60.0
) {
    init {
        require(sensorWindowSeconds in 3.0..10.0) {
            "sensorWindowSeconds 必须在 3–10 之间, 实际: $sensorWindowSeconds"
        }
    }
}

// ── 会话指标 ──────────────────────────────────────────────────────────────────

/**
 * 聚合会话指标，通过 [onMetricsUpdate] 回调投递给 UI。
 */
data class SessionMetrics(
    val cadence: CadenceSample? = null,
    val gait: GaitSnapshot? = null,
    val elapsedSeconds: Double = 0.0,
    val state: RunSessionState = RunSessionState.idle,
    val promptHistory: List<CoachPrompt> = emptyList()
)

// ── RunSessionManager ────────────────────────────────────────────────────────

/**
 * 实时跑步教练管线编排器:
 * `SensorCaptureManager → CadenceDetector → GaitAnalyzer → AudioCoachEngine`
 *
 * 状态机: `idle → ready → running → paused → stopped`
 *
 * 管理传感器捕获、步频检测、步态分析和语音教练。
 * 作为一个 [@Singleton] 存在，可以在多个 ViewModel 或服务中共享。
 *
 * 用法:
 * ```
 * // Hilt 自动注入
 * class MyViewModel @Inject constructor(
 *     private val sessionManager: RunSessionManager
 * ) : ViewModel() { ... }
 * ```
 */
@Singleton
class RunSessionManager @Inject constructor(
    @ApplicationContext private val appContext: Context
) {

    companion object {
        private const val TAG = "RunSessionManager"
        private const val METRICS_INTERVAL_MS = 1000L  // 1 Hz
        private const val MAX_PROMPT_HISTORY = 50
    }

    // ── 协程作用域 ────────────────────────────────────────────────────────────

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    // ── 公开属性 ──────────────────────────────────────────────────────────────

    /** 当前会话状态。 */
    private val _state = MutableStateFlow(RunSessionState.idle)
    val state: StateFlow<RunSessionState> = _state.asStateFlow()

    /** 会话配置。 */
    var config: RunSessionConfig = RunSessionConfig()
        private set

    /** 已用时间 (秒)，排除暂停时段。 */
    private val _elapsedSeconds = MutableStateFlow(0.0)
    val elapsedSeconds: StateFlow<Double> = _elapsedSeconds.asStateFlow()

    /** 本会话中播报过的教练提示历史。 */
    private val _promptHistory = MutableStateFlow<List<CoachPrompt>>(emptyList())
    val promptHistory: StateFlow<List<CoachPrompt>> = _promptHistory.asStateFlow()

    /** 聚合指标回调 (供 UI 更新)。 */
    var onMetricsUpdate: ((SessionMetrics) -> Unit)? = null

    /** 状态转换回调。 */
    var onStateChange: ((old: RunSessionState, new: RunSessionState) -> Unit)? = null

    /** 教练提示播报回调 (供 UI overlay)。 */
    var onCoachPrompt: ((CoachPrompt) -> Unit)? = null

    // ── 管线组件 ──────────────────────────────────────────────────────────────

    lateinit var sensorCaptureManager: SensorCaptureManager
        private set

    lateinit var cadenceDetector: CadenceDetector
        private set

    lateinit var gaitAnalyzer: GaitAnalyzer
        private set

    lateinit var audioCoach: AudioCoachEngine
        private set

    // ── 内部状态 ──────────────────────────────────────────────────────────────

    private var sessionStartTimeMillis: Long = 0L
    private var pauseStartTimeMillis: Long? = null
    private var totalPausedMillis: Long = 0L
    private var metricsJob: Job? = null
    private var sensorCollectJob: Job? = null

    // ── 初始化管线 ────────────────────────────────────────────────────────────

    /**
     * 使用默认配置或指定配置初始化管线组件。
     *
     * 一般在 [start] 之前调用。
     */
    fun configure(cfg: RunSessionConfig = RunSessionConfig()) {
        config = cfg

        sensorCaptureManager = SensorCaptureManager(appContext)

        cadenceDetector = CadenceDetector(
            alpha = 0.8,
            windowSeconds = config.sensorWindowSeconds,
            minCadenceSPM = 50.0,
            maxCadenceSPM = 240.0
        )

        gaitAnalyzer = GaitAnalyzer(
            windowSeconds = config.sensorWindowSeconds,
            samplingRate = config.samplingRate
        )

        audioCoach = AudioCoachEngine(
            appContext,
            language = config.coachingLanguage,
            minIntervalSeconds = config.minPromptIntervalSec
        )

        wirePipeline()
    }

    // ── 管线连接 ──────────────────────────────────────────────────────────────

    private fun wirePipeline() {
        // SensorCaptureManager → CadenceDetector + GaitAnalyzer
        sensorCollectJob?.cancel()
        sensorCollectJob = scope.launch {
            sensorCaptureManager.sensorFrames().collect { frame ->
                // 直接将完整帧喂给步频检测器 (含 timestampNanos 用于精确时间计算)
                cadenceDetector.process(frame)

                // 完整帧喂给步态分析器
                gaitAnalyzer.process(frame)
            }
        }

        // CadenceDetector → AudioCoach 评估
        scope.launch {
            cadenceDetector.currentCadence.collectLatest { cadence ->
                if (cadence != null && _state.value == RunSessionState.running) {
                    // Inject latest cadence into gait analyzer for snapshot enrichment
                    gaitAnalyzer.latestCadenceSPM = cadence.stepsPerMinute
                    val gait = gaitAnalyzer.currentSnapshot.value
                    audioCoach.evaluate(
                        cadence = cadence,
                        gait = gait,
                        targetCadence = config.targetCadenceSPM,
                        elapsedSeconds = _elapsedSeconds.value
                    )
                }
            }
        }

        // AudioCoach 提示 → UI 回调 + 历史
        audioCoach.onPromptQueued = { prompt ->
            val current = _promptHistory.value.toMutableList()
            current.add(prompt)
            if (current.size > MAX_PROMPT_HISTORY) {
                current.removeAt(0)
            }
            _promptHistory.value = current
            onCoachPrompt?.invoke(prompt)
        }
    }

    // ── 公开 API: 状态机 ──────────────────────────────────────────────────────

    /**
     * idle → ready → running。
     */
    fun start() {
        if (_state.value != RunSessionState.idle && _state.value != RunSessionState.ready) {
            Log.w(TAG, "无法从 ${_state.value} 启动")
            return
        }

        // 确保管线已配置
        if (!::sensorCaptureManager.isInitialized) {
            configure(config)
        }

        transitionTo(RunSessionState.ready)
        transitionTo(RunSessionState.running)

        sessionStartTimeMillis = System.currentTimeMillis()
        totalPausedMillis = 0L
        _elapsedSeconds.value = 0.0
        _promptHistory.value = emptyList()

        sensorCaptureManager.start()
        startMetricsTimer()
    }

    /** 暂停会话 (传感器继续运行，但指标和教练暂停)。 */
    fun pause() {
        if (_state.value != RunSessionState.running) return
        pauseStartTimeMillis = System.currentTimeMillis()
        transitionTo(RunSessionState.paused)
    }

    /** 从暂停恢复。 */
    fun resume() {
        if (_state.value != RunSessionState.paused) return
        pauseStartTimeMillis?.let { start ->
            totalPausedMillis += System.currentTimeMillis() - start
            pauseStartTimeMillis = null
        }
        transitionTo(RunSessionState.running)
    }

    /** 完全停止会话。 */
    fun stop() {
        if (_state.value == RunSessionState.idle || _state.value == RunSessionState.stopped) return

        sensorCaptureManager.stop()
        stopMetricsTimer()
        cadenceDetector.reset()
        gaitAnalyzer.reset()
        audioCoach.stopSpeaking()
        sensorCollectJob?.cancel()

        transitionTo(RunSessionState.stopped)
        transitionTo(RunSessionState.idle)
    }

    // ── 指标定时器 (~1 Hz) ────────────────────────────────────────────────────

    private fun startMetricsTimer() {
        metricsJob?.cancel()
        metricsJob = scope.launch {
            while (true) {
                delay(METRICS_INTERVAL_MS)
                if (_state.value != RunSessionState.running) continue

                // 更新已用时间 (扣除暂停)
                val now = System.currentTimeMillis()
                _elapsedSeconds.value =
                    ((now - sessionStartTimeMillis - totalPausedMillis) / 1000.0)
                    .coerceAtLeast(0.0)

                // 发射聚合指标
                val metrics = SessionMetrics(
                    cadence = cadenceDetector.currentCadence.value,
                    gait = gaitAnalyzer.currentSnapshot.value,
                    elapsedSeconds = _elapsedSeconds.value,
                    state = _state.value,
                    promptHistory = _promptHistory.value
                )
                onMetricsUpdate?.invoke(metrics)
            }
        }
    }

    private fun stopMetricsTimer() {
        metricsJob?.cancel()
        metricsJob = null
    }

    // ── 状态转换 ──────────────────────────────────────────────────────────────

    private fun transitionTo(newState: RunSessionState) {
        val oldState = _state.value
        if (oldState == newState) return
        _state.value = newState
        Log.i(TAG, "$oldState → $newState")
        onStateChange?.invoke(oldState, newState)
    }

    // ── 清理 ──────────────────────────────────────────────────────────────────

    /**
     * 显式销毁管线并释放资源。
     */
    fun shutdown() {
        stopMetricsTimer()
        sensorCollectJob?.cancel()
        sensorCaptureManager.stop()
        audioCoach.shutdown()
        Log.d(TAG, "RunSessionManager 已销毁")
    }
}
